### PR TITLE
Shifted sse loss to decision tree to streamline xgboost functionality and ease regularisation.

### PR DESCRIPTION
**Pull request by TirelessClock**
SSE_loss existed in xgboost directory beforehand. I shifted that to the Decision Trees directory and included that as a gain function to be used. 

All the xgboost loss functions are to be implemented at the decision tree level.

### CODE REVIEW

File: src/mlpack/methods/decision_tree/decision_tree.hpp [hunk 1]
```diff
@@ -15,19 +15,19 @@
...
 
-#include "gini_gain.hpp"
-#include "information_gain.hpp"
-#include "mad_gain.hpp"
-#include "mse_gain.hpp"
+#include "gain_functions/gini_gain.hpp"
+#include "gain_functions/information_gain.hpp"
+#include "gain_functions/mad_gain.hpp"
+#include "gain_functions/mse_gain.hpp"
+#include "gain_functions/sse_gain.hpp"
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree.hpp:22]: Many other subdirectories have files like `gain_functions/gain_functions.hpp` that simply include all the gain functions---do you want to do that here too?  It's just a convenience.

File: src/mlpack/methods/decision_tree/gain_functions/information_gain.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
- * @file methods/decision_tree/information_gain.hpp
+ * @file methods/decision_tree/gain_functions/information_gain.hpp
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/gain_functions/information_gain.hpp:2]: If you really want to go above and beyond you can update the header guards too 😄

File: src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp [hunk 3]
```diff
@@ -1,8 +1,8 @@
 /**
- * @file methods/xgboost/loss_functions/sse_loss.hpp
+ * @file methods/decision_tree/gain_functions/sse_gain.hpp
  * @author Rishabh Garg
  *
- * The sum of squared error loss class, which is a loss funtion for gradient
+ * The sum of squared error loss class, which is a loss function for gradient
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp:5]: Nice catch!

File: src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp [hunk 4]
```diff
@@ -25,13 +25,13 @@ namespace mlpack {
  *
  * Loss = 1 / 2 * (Observed - Predicted)^2
  */
-class SSELoss
+class SSEGain
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp:28]: Are you sure the equation above is correct?  I read the name change from loss to gain, and wanted to check whether the function actually was gain (highest value is best) or loss (lowest value is best), but then I noticed the sentence `We will try to minimize this value while training` (which would be loss), and then the equation---but the equation doesn't seem to match the implementation.  I didn't look into the paper to remember what exactly SSEGain (or loss) is; do you think you can double-check whether the comment is accurate?  (And it may be worth double-checking the implementation too!)

**Review by rcurtin — COMMENTED** [review]
Nice, we now have enough gain functions and splitting strategies that they need to be organized!  🚀  I just took a quick look and have some simple comments.  Did you accidentally include `log_gain.cpp` as part of this PR when you meant that for another PR?

File: src/mlpack/methods/decision_tree/decision_tree.hpp (hunk 1 above)
Comment (TirelessClock) [reply src/mlpack/methods/decision_tree/decision_tree.hpp:22]: Oh yeah, I'll do that now.

File: src/mlpack/methods/decision_tree/gain_functions/information_gain.hpp (hunk 2 above)
Comment (TirelessClock) [reply src/mlpack/methods/decision_tree/gain_functions/information_gain.hpp:2]: Ah sorry... I did that for most of them I guess I missed a few 🥴

File: src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp (hunk 4 above)
Comment (TirelessClock) [reply src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp:28]: I shifted this from a previous implementation in XGBoost, where it was called `SSE_loss`. I changed it to gain because the actual implementation was just a gain function in decision trees. I'll update the comments ASAP.

**Comment by TirelessClock** [comment]
> Nice, we now have enough gain functions and splitting strategies that they need to be organized! 🚀 I just took a quick look and have some simple comments. Did you accidentally include `log_gain.cpp` as part of this PR when you meant that for another PR?

@rcurtin That's also a part of this PR, I'm extending gain functions by adding new ones, more specific to xgboost. I put them in this PR because I thought it'd be better to bunch the gain function related work together. If you prefer I can put it in another PR.

**Comment by rcurtin** [comment]
Oh, sounds good, no worries (and no hurry).  Don't forget to add a test for the new gain function :+1: 

I don't think it should be a part of this PR but we will need to at some point document all the XGBoost-related functionalities somewhere in `doc/user/`, maybe by adding to `decision_tree.md`, maybe with a new `xgboost.md`, maybe both---it doesn't need to be done now, but I do want to make sure everything we add has some visibility in the documentation. :+1:

**Comment by TirelessClock** [comment]
Sure, I'll put it in another PR 👍
I'll update the documentation after I've finished a few more things in XGBoost. I guess this will fit better in the XGB documentation as it's specific to that.

**Review by zoq — COMMENTED** [review]
Looks good to me, waiting for the CI.

**Comment by TirelessClock** [comment]
@zoq Hey, I think it's ready to merge now. Please check.

**Comment by zoq** [comment]
Yes, can you resolve the merge conflicts? Let me know if you need help.

File: src/mlpack/methods/decision_tree/decision_tree.hpp [hunk 5]
```diff
@@ -15,20 +15,9 @@
...
-#include "best_binary_numeric_split.hpp"
-#include "random_binary_numeric_split.hpp"
-
-#include "best_binary_categorical_split.hpp"
-#include "all_categorical_split.hpp"
-
-#include "all_dimension_select.hpp"
-#include "random_dimension_select.hpp"
-#include "multiple_random_dimension_select.hpp"
+#include "gain_functions/gain_functions.hpp"
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree.hpp:18]: Since this is called `FitnessFunction` in [the documentation](https://www.mlpack.org/doc/user/methods/decision_tree.html#advanced-functionality-template-parameters), it probably makes sense to use the directory name `fitness_functions/` here.

File: src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp [hunk 6]
```diff
@@ -66,8 +66,15 @@ class SSELoss
    * @param begin The begin index to calculate gain.
    * @param end The end index to calculate gain.
    */
-  double Evaluate(const size_t begin, const size_t end)
+  template<bool UseWeights, typename MatType, typename WeightVecType>
+  double Evaluate(const MatType& input, 
+                  const WeightVecType& /* weights */,
+                  const size_t begin, 
+                  const size_t end)
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp:73]: Is this a `FitnessFunction`?  It doesn't seem to match the necessary API in the documentation.

Comment (rcurtin) [reply src/mlpack/methods/decision_tree/gain_functions/sse_gain.hpp:73]: Or is it only for [`DecisionTreeRegressor`](https://www.mlpack.org/doc/user/methods/decision_tree_regressor.html#advanced-functionality-template-parameters)?

**Review by zoq — COMMENTED** [review]
Currently this does not build, on the CI or locally. Because `sse_gain.hpp` is not part of the PR.

**Review by zoq — APPROVED** [review]
Okay from my side.

**Review by shrit — APPROVED** [review]
